baseplate/observers/metrics_tagged.py

```diff
@@ -0,0 +1,183 @@
+import logging
+
+from enum import Enum
+from typing import Any
+from typing import Dict
+from typing import List
+from typing import Optional
+
+from baseplate import _ExcInfo
+from baseplate import BaseplateObserver
+from baseplate import LocalSpan
+from baseplate import RequestContext
+from baseplate import Span
+from baseplate import SpanObserver
+from baseplate.lib import config
+from baseplate.lib import metrics
+from baseplate.observers.timeout import ServerTimeout
+
+
+logger = logging.getLogger(__name__)
+
+
+class Errors(Enum):
+    TIMED_OUT = 1
+    EXCEPTION = 2
+
+
+error_messages = {
+    Errors.TIMED_OUT: "timed_out",
+    Errors.EXCEPTION: "internal_server_error",
+}
+
+
+class TaggedMetricsBaseplateObserver(BaseplateObserver):
+    """Metrics collecting observer.
+
+    This observer reports metrics to statsd in the dogstatsd format. It does three important things:
+
+    * it tracks the time taken in serving each request.
+    * it batches all metrics generated during a request into as few packets
+      as possible.
+    * it adds tags to the metric if they are in the config tag whitelist
+
+    The batch is accessible to your application during requests as the
+    ``metrics`` attribute on the :py:class:`~baseplate.RequestContext`.
+
+    :param client: The client where metrics will be sent.
+    :param cfg: the parsed application config with the tag whitelist
+
+    """
+
+    def __init__(self, client: metrics.Client, whitelist: List[str]):
+        self.client = client
+        self.whitelist = whitelist
+
+    @classmethod
+    def from_config_and_client(
+        cls, raw_config: config.RawConfig, client: metrics.Client
+    ) -> "TaggedMetricsBaseplateObserver":
+        cfg = config.parse_config(
+            raw_config,
+            {
+                "metrics": {
+                    "whitelist": config.Optional(
+                        config.TupleOf(config.String), default=["endpoint", "success", "error"]
+                    ),
+                }
+            },
+        )
+        return cls(client, whitelist=cfg.metrics.whitelist)
+
+    def on_server_span_created(self, context: RequestContext, server_span: Span) -> None:
+        batch = self.client.batch()
+        context.metrics = batch
+        observer = TaggedMetricsServerSpanObserver(batch, server_span, self.whitelist)
+        server_span.register(observer)
+
+
+class TaggedMetricsServerSpanObserver(SpanObserver):
+    def __init__(self, batch: metrics.Batch, server_span: Span, whitelist: list):
+        self.batch = batch
+        self.span = server_span
+        self.base_name = "baseplate.server"
+        self.whitelist = whitelist
+        self.tags: Dict[str, str] = {}
+        self.timer = batch.timer(self.base_name)
+
+    def on_start(self) -> None:
+        self.span.set_tag("endpoint", self.span.name)
```

Comment:
```suggestion
        self.tags["endpoint"] = self.span.name
```

another instance of not wanting to call set_tag() on the abstract span from within an observer